R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+#' environment are converted back to standard tags objects with `$asTags()`.
+#'
+#' Tag environments also contain an extra field, `.$parent`. The `.$parent`
+#' value contains their parent tag environment. The top level tags supplied to
+#' `tagQuery()` will also have a shared parent element. (The shared parent
+#' element will have a `NULL` `.$parent` value.) This allows for performing
+#' sibling alterations at the top level of the tag query object.
+#'
+#' The set of tag environments in a pointing to each other within a tag query
+#' object can be thought of as a linked list while allowing for a "1 to many"
+#' parent to child relationship and up to 1 parent per child.
+#'
+#' ## Tag Query
+#'
+#' A `tagQuery()` behaves simliar to an R6 object (but a `tagQuery()` object is
+#' not implemented with `R6`). The `tagQuery()`'s methods will return itself as
+#' much as possible, unless the method is directly asking for information, e.g.
+#' `$selected()` or `$asTags()`.
+#'
+#' Internally, two important pieces of information are maintained: the root
+#' elements and the selected elements. The root tag environment will always
+#' point (after upgrading to a tag environment) to the original tag object
+#' provided to `tagQuery(tag=)`. However, the selected elements are a list of
+#' tag environments that update for every `$find(cssSelector)` call.  The
+#' selected elements are initialized to a list containing the `root` tag
+#' environment. All `tagQuery()` methods will act on the selected elements
+#' unless declared otherwise.
+#'
+#'
+#' @section Limitations:
+#'
+#'   `tagQuery()`s can **not** be used directly within typical `tag` locations.
+#'   An error should be thrown. Instead, please call `$asTags()` to retrieve the
+#'   tag structures of the selected tag elements or root element respectively.
+#'
+#' @param tags Any standard tag object or `tagList()`. If a `list()` or
+#'   `tagList()` is provided, a `tagList()` will be returned when calling
+#'   `$asTags()`.
+#' @return A `tagQuery()` object. The `tag` supplied will be considered the
+#'   `root` object. At the time of initialization, the `root` is also considered
+#'   the single selected item. If any selections are made, the selected elements
+#'   will be updated.
+#' @md
+#' @export
+tagQuery <- function(tags) {
+
+  tags <- flattenTagsRaw(tags %||% list())
+  tags <- asTagEnv(
+    wrapWithRootTag(tags)
+  )
+
+  selected <- tagQueryFindReset(tags)
+
+  tagQuery_(tags, selected)
+}
+
+#' @rdname tagQuery
+#' @aliases NULL
+#' @usage NULL
+#' @md
+tagQuery_ <- function(root, selected) {
+  if (!isTagEnv(root)) {
+    stop("`tagQuery_(root=)` must be a tag environment")
+  }
+
+  # Use `var_` names to avoid namespace collision
+  # Make sure all elements are tag envs
+  rebuild_ <- function() {
+    # safe to do as `root` will never be turned into a standard list
+    asTagEnv(root)
+  }
+  selected_ <- selected
```

Comment:
I think it would be slightly cleaner to just take `selected_` as an argument to the `tagQuery_` function. Then if `selected_` is changed, we don't hang on to a reference of the original object, and there's also no confusion about the `selected` argument used in the internal functions.